prototypes/palloc/pool_alloc.c

```diff
@@ -0,0 +1,212 @@
+        (pool -> pp_destructor) (curr_pi -> pi_data, (pool -> pp_itemsize));
+    }
+    free(slab);
+}
+
+void *slab_alloc(pool_slab_t *slab, size_t size) {
+    uint8_t *bitmap = slab -> ph_bitmap;
+    int i = 0;
+    while (bitmap[i]) { //TODO: bitmap improvements
+        i++;
+    }
+    pool_item_t *found_pi = (pool_item_t*) ((uint64_t) slab+(slab -> ph_start)+i*(size+sizeof(pool_item_t)));
+    if (found_pi -> pi_guard_number != 0xcafebabecafebabe) {
+        panic("memory corruption at item 0x%lx\n", (uint64_t) found_pi);
+    }
+    (slab -> ph_nused)++;
+    (slab -> ph_nfree)--;
+    bitmap[i] = 1;
+    DEBUG_PRINT("Allocated an item (0x%lx) at slab 0x%lx, index %d\n", (uint64_t) found_pi -> pi_data, (uint64_t) found_pi -> pi_slab, i);
+    return found_pi -> pi_data;
+} 
+
+void pool_init(pool_t *pool, size_t size, 
+        void (*constructor)(void *, size_t), void (*destructor)(void *, size_t)) {
+    /*
+    init object and give it 1-page cache;
+    */
+    DEBUG_PRINT("Entering pool_init\n");
+    size = align_to_binary_pow(size, 3); //Align to 64-bit (8 byte) word
+    LIST_INIT(&(pool -> pp_empty_slabs));
+    LIST_INIT(&(pool -> pp_full_slabs));
+    LIST_INIT(&(pool -> pp_part_slabs));
+    pool_slab_t *first_slab = create_slab(size, constructor);
+    LIST_INSERT_HEAD(&(pool -> pp_empty_slabs), first_slab, ph_slablist);
+    pool -> pp_constructor = constructor;
+    pool -> pp_destructor = destructor;
+    pool -> pp_itemsize = size;
+    pool -> pp_nslabs = 1;
+    pool -> pp_nitems = first_slab -> ph_ntotal;
+    DEBUG_PRINT("Initialized new pool at 0x%lx\n", (uint64_t) pool);
+}
+
+
+void pool_destroy(pool_t* pool) {
+    /*
+    take some objects from the cache;
+    destroy the objects;
+    free the underlying memory;
+    */
+
+    DEBUG_PRINT("Entering pool_destroy\n");
+
+    uint64_t *tmp = (uint64_t*) pool;
+
+    for (size_t i = 0; i < sizeof(pool_t)/8; i++) {
+        if (tmp[i] == 0xdeadbeefdeadbeef) {
+            panic("double free at pool 0x%lx\n", (uint64_t) pool);
+        }
+    }
+
+    DEBUG_PRINT("Destroying empty slabs\n");
+    pool_slab_t *it = LIST_FIRST(&(pool -> pp_empty_slabs));
+    while(it) {
+        pool_slab_t *next = LIST_NEXT(it, ph_slablist);
+        LIST_REMOVE(it, ph_slablist);
+        destroy_slab(it, pool);
+        it = next;
+    }
+
+    DEBUG_PRINT("Destroying partially filled slabs\n");
+    it = LIST_FIRST(&(pool -> pp_part_slabs));
+    while(it) {
+        pool_slab_t *next = LIST_NEXT(it, ph_slablist);
+        LIST_REMOVE(it, ph_slablist);
+        destroy_slab(it, pool);
+        it = next;
+    }
+
+    DEBUG_PRINT("Destroying full slabs\n");
+    it = LIST_FIRST(&(pool -> pp_full_slabs));
+    while(it) {
+        pool_slab_t *next = LIST_NEXT(it, ph_slablist);
+        LIST_REMOVE(it, ph_slablist);
+        destroy_slab(it, pool);
+        it = next;
+    }
+
+    for (size_t i = 0; i < sizeof(pool_t)/8; i++) {
+        tmp[i] = 0xdeadbeefdeadbeef;
```

Comment:
Am I correct that this causes writes to memory regions that were already freed? It might be better to first perform these writes (e.g. while checking the `tmp[i]` values at the beginning of `pool_destroy`), and destroy slabs afterwards.